src/lsptoolshost/profiling/profiling.ts

```diff
@@ -24,3 +30,248 @@ export function getProfilingEnvVars(outputChannel: LogOutputChannel): NodeJS.Pro
+                }
+            );
+        })
+    );
+}
+
+async function executeDotNetTraceCommand(
+    languageServer: RoslynLanguageServer,
+    progress: vscode.Progress<{
+        message?: string;
+        increment?: number;
+    }>,
+    outputChannel: vscode.LogOutputChannel,
+    cancellationToken: vscode.CancellationToken
+): Promise<void> {
+    const processId = languageServer.processId;
+
+    if (!processId) {
+        throw new Error(vscode.l10n.t('Language server process not found, ensure the server is running.'));
+    }
+
+    let traceFolder: string | undefined = '';
+    if (vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders?.length >= 1) {
+        traceFolder = vscode.workspace.workspaceFolders[0].uri.fsPath;
+    }
+
+    // Prompt the user for the folder to save the trace file
+    // By default, choose the first workspace folder if available.
+    const uris = await vscode.window.showOpenDialog({
+        canSelectFiles: false,
+        canSelectFolders: true,
+        canSelectMany: false,
+        defaultUri: traceFolder ? vscode.Uri.file(traceFolder) : undefined,
+        openLabel: vscode.l10n.t('Select Trace Folder'),
+        title: vscode.l10n.t('Select Folder to Save Trace File'),
+    });
+
+    if (uris === undefined || uris.length === 0) {
+        // User cancelled the dialog
+        return;
+    }
+
+    traceFolder = uris[0].fsPath;
+
+    if (!fs.existsSync(traceFolder)) {
+        throw new Error(vscode.l10n.t(`Folder for trace file {0} does not exist`, traceFolder));
+    }
+
+    const dotnetTraceArgs = `--process-id ${processId} --clreventlevel informational --providers "Microsoft-DotNETCore-SampleProfiler,Microsoft-Windows-DotNETRuntime"`;
+
+    // Show an input box pre-populated with the default dotnet trace arguments
+    const userArgs = await vscode.window.showInputBox({
+        value: dotnetTraceArgs,
+        placeHolder: vscode.l10n.t('Enter dotnet-trace arguments'),
+        prompt: vscode.l10n.t('You can modify the default dotnet-trace arguments if needed'),
+    });
+    if (userArgs === undefined) {
+        // User cancelled the input box
+        return;
+    }
+
+    const terminal = await getOrCreateTerminal(traceFolder, outputChannel);
+
+    const dotnetTraceInstalled = await verifyOrAcquireDotnetTrace(traceFolder, progress, outputChannel);
+    if (!dotnetTraceInstalled) {
+        // Cancelled or unable to install dotnet-trace
+        return;
+    }
+
+    const args = ['collect', ...userArgs.split(' ')];
+
+    progress.report({ message: vscode.l10n.t('Recording trace...') });
+    await runDotnetTrace(args, terminal, cancellationToken);
+}
+
+async function verifyOrAcquireDotnetTrace(
+    folder: string,
+    progress: vscode.Progress<{
+        message?: string;
+        increment?: number;
+    }>,
+    channel: vscode.LogOutputChannel
+): Promise<boolean> {
+    try {
+        await execChildProcess('dotnet-trace --version', folder, process.env);
```

Comment:
it seems relatively common that we can't get the shell execution for a terminal (which means we can't observe command output).  For example, in a windows sandbox this appears to be the case.

So instead of fiddling with the terminal, for simple commands where we must have the output, we can just run them directly.